src/NuGet.Core/NuGet.Packaging/Signing/SignVerifier.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Signing
+{
+    /// <summary>
+    /// Loads trust providers and verifies package signatures.
+    /// </summary>
+    public class SignVerifier
+    {
+        private readonly List<ISignTrustProvider> _trustProviders;
+        private readonly SignVerifierSettings _settings;
+
+        public SignVerifier(IEnumerable<ISignTrustProvider> trustProviders, SignVerifierSettings settings)
+        {
+            _trustProviders = trustProviders?.ToList() ?? throw new ArgumentNullException(nameof(trustProviders));
+            _settings = settings ?? throw new ArgumentNullException(nameof(settings));
+        }
+
+        public async Task<VerifySignaturesResult> VerifySignaturesAsync(ISignPackageReader package, ILogger logger, CancellationToken token)
```

Comment:
we also need to validate the manifest and the hash for the manifest in the signature or the hash of the signature file for a counter signature.

Reply:
not part of this PR